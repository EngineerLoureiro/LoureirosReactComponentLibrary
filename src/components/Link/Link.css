.link {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 6px;
  border-radius: 4px;
  font-family: "Noto Sans";
  font-weight: 500;
  font-size: 16px;
  line-height: 24px;
  color: #525252;
}

.link:hover {
  color: #171717;
}

.link:focus {
  background-color: #ffffff;
  box-shadow: 0px 0px 0px 4px
    rgb(0.2666666805744171 0.2980392277240753 0.9058823585510254 / 0.12);
}

.link--disabled {
  color: #a3a3a3;
  cursor: not-allowed;
}
