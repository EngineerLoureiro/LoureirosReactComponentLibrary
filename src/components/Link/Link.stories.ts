// Replace your-framework with the framework you are using, e.g. react-vite, nextjs, nextjs-vite, etc.
import type { Meta, StoryObj } from "@storybook/react-vite";

import { Link } from "./Link";

import { fn } from "storybook/test";

//ðŸ‘‡ This default export determines where your story goes in the story list
const meta = {
  title: "Loureiro'sComponentLibrary/Link",
  component: Link,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  args: { onClick: fn(() => alert("Link clicked")) },
} satisfies Meta<typeof Link>;

export default meta;
type Story = StoryObj<typeof meta>;

export const ActiveLink: Story = {
  args: {
    label: "Click Me",
  },
};

export const DisabledLink: Story = {
  args: {
    label: "Click Me",
    disabled: true,
  },
};
