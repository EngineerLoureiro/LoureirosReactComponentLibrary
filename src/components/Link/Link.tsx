import classNames from "classnames";
import { AnchorHTMLAttributes } from "react";
import "./Link.css";

export interface LinkProps extends AnchorHTMLAttributes<HTMLAnchorElement> {
  label: string;
  disabled?: boolean;
}

export function Link({
  target,
  rel,
  label,
  className,
  disabled,
  onClick,
  ...rest
}: LinkProps) {
  const handleOnClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
    if (disabled) {
      e.preventDefault();
      e.stopPropagation();
      return;
    } else if (onClick) {
      onClick(e);
    }
  };
  return (
    <a
      target={target}
      rel={target === "_blank" ? "noopener noreferrer" : rel}
      className={classNames("link", { "link--disabled": disabled }, className)}
      onClick={handleOnClick}
      aria-disabled={disabled}
      tabIndex={disabled ? -1 : rest.tabIndex}
      {...rest}
    >
      {label}
    </a>
  );
}
