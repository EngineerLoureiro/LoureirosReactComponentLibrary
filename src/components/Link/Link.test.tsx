import { render, screen, fireEvent } from "@testing-library/react";
import { Link } from "./Link";

describe("Link component", () => {
  it("renders with correct label and href", () => {
    render(<Link href="/about" label="About" />);
    const link = screen.getByRole("link", { name: "About" });
    expect(link).toHaveAttribute("href", "/about");
  });

  it("adds rel = 'noopener noreferrer' when target='_blank'", () => {
    render(<Link href="/about" label="About" target="_blank" />);
    const link = screen.getByRole("link", { name: "About" });
    expect(link).toHaveAttribute("rel", "noopener noreferrer");
  });

  it("adds rel ='nofollow ", () => {
    render(<Link href="/about" label="About" rel="nofollow" />);
    const link = screen.getByRole("link", { name: "About" });
    expect(link).toHaveAttribute("rel", "nofollow");
  });

  it("calls onClick when clicked", () => {
    const handleOnClick = jest.fn();
    render(<Link href="/about" label="About" onClick={handleOnClick} />);
    fireEvent.click(screen.getByRole("link", { name: "About" }));
    expect(handleOnClick).toHaveBeenCalled();
  });
  it("disables the link", () => {
    const handleOnClick = jest.fn();
    render(
      <Link
        href="/about"
        label="About"
        onClick={handleOnClick}
        disabled={true}
      />
    );
    const link = screen.getByRole("link", { name: "About" });
    fireEvent.click(link);
    expect(handleOnClick).toHaveBeenCalledTimes(0);
    expect(link).toHaveAttribute("aria-disabled", "true");
    expect(link).toHaveAttribute("tabindex", "-1");
  });

  it("renders valid Link", () => {
    const { asFragment } = render(<Link href="/about" label="About" />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders disabled Link", () => {
    const { asFragment } = render(
      <Link href="/about" label="About" disabled={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
