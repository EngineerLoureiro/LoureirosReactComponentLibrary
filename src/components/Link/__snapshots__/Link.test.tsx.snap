// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component renders disabled Link 1`] = `
<DocumentFragment>
  <a
    aria-disabled="true"
    class="link link--disabled"
    href="/about"
    tabindex="-1"
  >
    About
  </a>
</DocumentFragment>
`;

exports[`Link component renders valid Link 1`] = `
<DocumentFragment>
  <a
    class="link"
    href="/about"
  >
    About
  </a>
</DocumentFragment>
`;
